### PR TITLE
ENA Status Sync commands

### PR DESCRIPTION
**Pull request by mberacochea**
This is a port of the code that syncs the status of the Studies, Samples, Runs and Assemblies from ENA into the API.
Original code -> https://github.com/EBI-Metagenomics/mi-automation/blob/develop/legacy_production/tools/production/emg-object-status-checker.py

I've tried to simplify the rules about how we sync the status. There is some duplicated code, so it's possible to generalize  the commands into a single one. I choose not to, so we can have a clearer picture of how the code works.

Pending: Unit tests for the assembly-sync command.

### CODE REVIEW

File: emgena/management/commands/sync_runs_with_ena.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        logger.info("Starting...")
+
+        offset = 0
+        batch_size = 1000
+        runs_count = emg_models.Run.objects.count()
+
+        logger.info(f"Total Runs on EMG {runs_count}")
+
+        while offset < runs_count:
+            emg_runs_batch = emg_models.Run.objects.all()[offset:batch_size]
```
Comment (SandyRogers) [inline emgena/management/commands/sync_runs_with_ena.py:43]:
```suggestion
            emg_runs_batch = emg_models.Run.objects.all()[offset:offset + batch_size]
```

File: emgena/management/commands/sync_runs_with_ena.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            ena_runs_batch = ena_models.Run.objects.filter(
+                run_id__in=[run.accession for run in emg_runs_batch]
+            )
+
+            for emg_run in emg_runs_batch:
+                ena_run = next(
+                    (el for el in ena_runs_batch if el.run_id == emg_run.accession),
+                    None,
+                )
```
Comment (SandyRogers) [inline emgena/management/commands/sync_runs_with_ena.py:52]: Does matching the ENA runs like this actually save any ENA queries? Isn't the `ena_runs_batch` query lazily executed anyway, so each iterator will do a new query anyway?

File: emgena/management/commands/sync_samples_with_ena.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        samples_count = emg_models.Sample.objects.count()
+
+        logger.info(f"Total Samples on EMG {samples_count}")
+
+        while offset < samples_count:
+            # TODO: review this rule, I didn't have enought time to review it.
+            # ported from: https://github.com/EBI-Metagenomics/mi-automation/blob/develop/legacy_production/tools/production/emg-object-status-checker.py#L242
+            emg_samples_batch = emg_models.Sample.objects.exclude(
+                accession__startswith="GCA_"
+            )[offset:batch_size]
```
Comment (SandyRogers) [inline emgena/management/commands/sync_samples_with_ena.py:47]:
```suggestion
            )[offset:offset + batch_size]
```

**Review by SandyRogers — CHANGES_REQUESTED** [review]
Nice work :) 
A few things look like bugs to me from reading the code (not running it). If I've misunderstood and they are actually correct as is, maybe let's drop some comments in there as to why... so I don't "fix" them in future!

File: emgena/management/commands/sync_assemblies_with_ena.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                if ena_assembly.status is None:
+                    logger.error(f"{emg_assembly} on ENA has no value on the column status.")
+                    continue
+
+                emg_assembly.sync_with_ena_status(ena_run.status)
+
+            emg_models.Assembly.objects.bulk_update(
+                emg_runs_batch, ["is_private", "is_suppressed", "reason"]
+            )
+            logger.info(f"Batch {assemblies_count / batch_size::.0f} processed.")
```
[the remark this answers is not in the document]
Comment (mberacochea) [reply emgena/management/commands/sync_assemblies_with_ena.py:66]: yeah... I forgot to clean that statement. Good catch

File: emgapi/models.py [hunk 5]
```diff
@@ -68,12 +78,78 @@ class Reason(models.IntegerChoices):
...
     suppresion_reason = models.IntegerField(db_column='REASON', blank=True, null=True, choices=Reason.choices)
 
-    def suppress(self, reason=None):
+    def suppress(self, suppresion_reason=None, save=True):
         self.is_suppressed = True
         self.suppressed_at = timezone.now()
-        if reason:
-            self.suppresion_reason = reason
-        self.save()
+        self.suppresion_reason = suppresion_reason
```
[the remark this answers is not in the document]
Comment (mberacochea) [reply emgapi/models.py:84]: Definitely, I'll fix that typo too

File: emgapi/models.py [hunk 6]
```diff
@@ -68,12 +78,78 @@ class Reason(models.IntegerChoices):
...
+        if (
+            ena_model_status
+            in [
+                ENAStatus.SUPPRESSED,
+                ENAStatus.KILLED,
+                ENAStatus.TEMPORARY_SUPPRESSED,
+                ENAStatus.TEMPORARY_KILLED,
+                ENAStatus.CANCELLED,
+            ]
+            and not self.is_suppressed
```
[the remark this answers is not in the document]
Comment (mberacochea) [reply emgapi/models.py:129]: I don't think that is an issue, the legacy code was doing that AFAIK... so I didn't won't to change that

File: emgena/management/commands/sync_runs_with_ena.py (hunk 2 above)
Comment (mberacochea) [reply emgena/management/commands/sync_runs_with_ena.py:52]: I need to check that...

Comment (mberacochea) [reply emgena/management/commands/sync_runs_with_ena.py:52]: I've just added a list() to force the evaluation of the QS

**Review by KeteSakharova — COMMENTED** [review]
I've read carefully

**Pull request merged**